File: pipeline/adapters/libnova_adapter/src/ephemerides.c
#include "common.h"

/* ------------------------------------------------------------------ */
/* Experiment: solar_position                                          */
/* Sun geocentric RA/Dec via libnova VSOP87                            */
/* Input per line: jd_tt                                               */
/* ------------------------------------------------------------------ */

void run_solar_position(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    printf("{\"experiment\":\"solar_position\",\"library\":\"libnova\",");
    printf("\"model\":\"libnova_VSOP87\",");
    printf("\"count\":%d,\"cases\":[\n", n);

    for (int i = 0; i < n; i++) {
        double jd_tt;
        if (scanf("%lf", &jd_tt) != 1) {
            fprintf(stderr, "bad input line %d\n", i); exit(1);
        }

        struct ln_equ_posn posn;
        ln_get_solar_equ_coords(jd_tt, &posn);

        /* libnova: ra in degrees (0..360), dec in degrees */
        double ra_rad  = posn.ra  * (M_PI / 180.0);
        double dec_rad = posn.dec * (M_PI / 180.0);

        /* Get distance */
        double dist_au = ln_get_earth_solar_dist(jd_tt);

        if (i > 0) printf(",\n");
        printf("{\"jd_tt\":%.15f,", jd_tt);
        printf("\"ra_rad\":%.17e,\"dec_rad\":%.17e,\"dist_au\":%.17e}", ra_rad, dec_rad, dist_au);
    }
    printf("\n]}\n");
}

/* ------------------------------------------------------------------ */
/* Experiment: lunar_position                                          */
/* Moon geocentric RA/Dec via libnova ELP 2000-82B                     */
/* Input per line: jd_tt                                               */
/* ------------------------------------------------------------------ */

void run_lunar_position(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    printf("{\"experiment\":\"lunar_position\",\"library\":\"libnova\",");
    printf("\"model\":\"libnova_ELP2000\",");
    printf("\"count\":%d,\"cases\":[\n", n);

    for (int i = 0; i < n; i++) {
        double jd_tt;
        if (scanf("%lf", &jd_tt) != 1) {
            fprintf(stderr, "bad input line %d\n", i); exit(1);
        }

        struct ln_equ_posn posn;
        ln_get_lunar_equ_coords(jd_tt, &posn);

        double ra_rad  = posn.ra  * (M_PI / 180.0);
        double dec_rad = posn.dec * (M_PI / 180.0);

        double dist_km = ln_get_lunar_earth_dist(jd_tt);

        if (i > 0) printf(",\n");
        printf("{\"jd_tt\":%.15f,", jd_tt);
        printf("\"ra_rad\":%.17e,\"dec_rad\":%.17e,\"dist_km\":%.17e}", ra_rad, dec_rad, dist_km);
    }
    printf("\n]}\n");
}

void run_planet_position(const char *experiment,
                                const char *model,
                                planet_equ_fn_t equ_fn,
                                planet_dist_fn_t dist_fn) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    printf("{\"experiment\":\"%s\",\"library\":\"libnova\",", experiment);
    printf("\"model\":\"%s\",", model);
    printf("\"count\":%d,\"cases\":[\n", n);

    for (int i = 0; i < n; i++) {
        double jd_tt;
        if (scanf("%lf", &jd_tt) != 1) {
            fprintf(stderr, "bad input line %d\n", i); exit(1);
        }

        struct ln_equ_posn posn;
        equ_fn(jd_tt, &posn);

        double ra_rad = posn.ra * (M_PI / 180.0);
        double dec_rad = posn.dec * (M_PI / 180.0);
        double dist_au = dist_fn(jd_tt);

        if (i > 0) printf(",\n");
        printf("{\"jd_tt\":%.15f,", jd_tt);
        printf("\"ra_rad\":%.17e,\"dec_rad\":%.17e,\"dist_au\":%.17e}", ra_rad, dec_rad, dist_au);
    }
    printf("\n]}\n");
}

/* ------------------------------------------------------------------ */
/* Geometric ephemeris runners                                         */
/*                                                                     */
/* These honour the LAB_LANE=geometric contract: no nutation,          */
/* no aberration, no light-time correction. They use libnova's         */
/* rectangular helio/geo helpers, which return AU vectors aligned      */
/* with equatorial J2000 (≈ICRF) via the J2000 obliquity rotation in  */
/* ln_get_rect_from_helio (sin_e=0.397777156, cos_e=0.917482062).      */
/* Output frame: equatorial J2000 / ICRF, comparable directly to JPL   */
/* Horizons VECTORS (VEC_CORR=NONE, REF_PLANE=FRAME, REF_SYSTEM=ICRF). */
/* ------------------------------------------------------------------ */

static void rect_to_radec_au(double X, double Y, double Z,
                             double *ra_rad, double *dec_rad, double *dist_au) {
    double r = sqrt(X * X + Y * Y + Z * Z);
    double ra = atan2(Y, X);
    if (ra < 0.0) ra += 2.0 * M_PI;
    *ra_rad = ra;
    *dec_rad = asin(Z / r);
    *dist_au = r;
}

void run_solar_position_geometric(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    printf("{\"experiment\":\"solar_position\",\"library\":\"libnova\",");
    printf("\"model\":\"libnova_VSOP87_geometric\",");
    printf("\"count\":%d,\"cases\":[\n", n);

    for (int i = 0; i < n; i++) {
        double jd_tt;
        if (scanf("%lf", &jd_tt) != 1) {
            fprintf(stderr, "bad input line %d\n", i); exit(1);
        }
        struct ln_rect_posn rect;
        ln_get_solar_geo_coords(jd_tt, &rect);
        double ra, dec, dist;
        rect_to_radec_au(rect.X, rect.Y, rect.Z, &ra, &dec, &dist);
        if (i > 0) printf(",\n");
        printf("{\"jd_tt\":%.15f,", jd_tt);
        printf("\"ra_rad\":%.17e,\"dec_rad\":%.17e,\"dist_au\":%.17e}", ra, dec, dist);
    }
    printf("\n]}\n");
}

/* J2000 obliquity (degrees → sin/cos) for the ecliptic→equatorial rotation
 * applied to the lunar geocentric vector, which comes out of libnova in the
 * inertial mean ecliptic and equinox of J2000. */
#define LN_J2000_SIN_E 0.397777156
#define LN_J2000_COS_E 0.917482062

void run_lunar_position_geometric(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    printf("{\"experiment\":\"lunar_position\",\"library\":\"libnova\",");
    printf("\"model\":\"libnova_ELP2000_geometric\",");
    printf("\"count\":%d,\"cases\":[\n", n);

    for (int i = 0; i < n; i++) {
        double jd_tt;
        if (scanf("%lf", &jd_tt) != 1) {
            fprintf(stderr, "bad input line %d\n", i); exit(1);
        }
        struct ln_rect_posn ecl;
        /* precision=0 → highest accuracy ELP 2000-82B series */
        ln_get_lunar_geo_posn(jd_tt, &ecl, 0.0);
        /* Rotate ecliptic-J2000 → equatorial-J2000 (≈ICRF). */
        double X = ecl.X;
        double Y = ecl.Y * LN_J2000_COS_E - ecl.Z * LN_J2000_SIN_E;
        double Z = ecl.Y * LN_J2000_SIN_E + ecl.Z * LN_J2000_COS_E;
        /* Lunar coords are in km already, so keep dist_km. */
        double r_km = sqrt(X * X + Y * Y + Z * Z);
        double ra = atan2(Y, X); if (ra < 0.0) ra += 2.0 * M_PI;
        double dec = asin(Z / r_km);
        if (i > 0) printf(",\n");
        printf("{\"jd_tt\":%.15f,", jd_tt);
        printf("\"ra_rad\":%.17e,\"dec_rad\":%.17e,\"dist_km\":%.17e}", ra, dec, r_km);
    }
    printf("\n]}\n");
}

void run_planet_position_geometric(const char *experiment,
                                          const char *model,
                                          planet_rect_helio_fn_t rect_helio_fn) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    printf("{\"experiment\":\"%s\",\"library\":\"libnova\",", experiment);
    printf("\"model\":\"%s\",", model);
    printf("\"count\":%d,\"cases\":[\n", n);

    for (int i = 0; i < n; i++) {
        double jd_tt;
        if (scanf("%lf", &jd_tt) != 1) {
            fprintf(stderr, "bad input line %d\n", i); exit(1);
        }
        /* Heliocentric planet & Earth, both in equatorial-J2000 AU rect coords.
         * Geocentric body = planet − Earth (no light-time, no aberration). */
        struct ln_rect_posn p, e;
        rect_helio_fn(jd_tt, &p);
        ln_get_earth_rect_helio(jd_tt, &e);
        double ra, dec, dist;
        rect_to_radec_au(p.X - e.X, p.Y - e.Y, p.Z - e.Z, &ra, &dec, &dist);
        if (i > 0) printf(",\n");
        printf("{\"jd_tt\":%.15f,", jd_tt);
        printf("\"ra_rad\":%.17e,\"dec_rad\":%.17e,\"dist_au\":%.17e}", ra, dec, dist);
    }
    printf("\n]}\n");
}

void run_solar_position_geometric_perf(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }
    double *jds = malloc(n * sizeof(double));
    for (int i = 0; i < n; i++) {
        if (scanf("%lf", &jds[i]) != 1) { fprintf(stderr, "bad input line %d\n", i); exit(1); }
    }
    for (int i = 0; i < n && i < 100; i++) {
        struct ln_rect_posn r;
        ln_get_solar_geo_coords(jds[i], &r);
        (void)r;
    }
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    double sink = 0.0;
    for (int i = 0; i < n; i++) {
        struct ln_rect_posn r;
        ln_get_solar_geo_coords(jds[i], &r);
        sink += r.X + r.Y + r.Z;
    }
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed_ns = (t1.tv_sec - t0.tv_sec) * 1e9 + (t1.tv_nsec - t0.tv_nsec);
    double per_op_ns = elapsed_ns / n;
    printf("{\"experiment\":\"solar_position_perf\",\"library\":\"libnova\",");
    printf("\"count\":%d,\"total_ns\":%.0f,\"per_op_ns\":%.1f,", n, elapsed_ns, per_op_ns);
    printf("\"throughput_ops_s\":%.0f,\"_sink\":%.17e}\n",
           (double)n / (elapsed_ns * 1e-9), sink);
    free(jds);
}

void run_lunar_position_geometric_perf(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }
    double *jds = malloc(n * sizeof(double));
    for (int i = 0; i < n; i++) {
        if (scanf("%lf", &jds[i]) != 1) { fprintf(stderr, "bad input line %d\n", i); exit(1); }
    }
    for (int i = 0; i < n && i < 100; i++) {
        struct ln_rect_posn r;
        ln_get_lunar_geo_posn(jds[i], &r, 0.0);
        (void)r;
    }
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    double sink = 0.0;
    for (int i = 0; i < n; i++) {
        struct ln_rect_posn r;
        ln_get_lunar_geo_posn(jds[i], &r, 0.0);
        sink += r.X + r.Y + r.Z;
    }
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed_ns = (t1.tv_sec - t0.tv_sec) * 1e9 + (t1.tv_nsec - t0.tv_nsec);
    double per_op_ns = elapsed_ns / n;
    printf("{\"experiment\":\"lunar_position_perf\",\"library\":\"libnova\",");
    printf("\"count\":%d,\"total_ns\":%.0f,\"per_op_ns\":%.1f,", n, elapsed_ns, per_op_ns);
    printf("\"throughput_ops_s\":%.0f,\"_sink\":%.17e}\n",
           (double)n / (elapsed_ns * 1e-9), sink);
    free(jds);
}

void run_planet_position_geometric_perf(const char *experiment,
                                               planet_rect_helio_fn_t rect_helio_fn) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }
    double *jds = malloc(n * sizeof(double));
    for (int i = 0; i < n; i++) {
        if (scanf("%lf", &jds[i]) != 1) { fprintf(stderr, "bad input line %d\n", i); exit(1); }
    }
    for (int i = 0; i < n && i < 100; i++) {
        struct ln_rect_posn p, e;
        rect_helio_fn(jds[i], &p);
        ln_get_earth_rect_helio(jds[i], &e);
        (void)p; (void)e;
    }
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    double sink = 0.0;
    for (int i = 0; i < n; i++) {
        struct ln_rect_posn p, e;
        rect_helio_fn(jds[i], &p);
        ln_get_earth_rect_helio(jds[i], &e);
        sink += (p.X - e.X) + (p.Y - e.Y) + (p.Z - e.Z);
    }
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed_ns = (t1.tv_sec - t0.tv_sec) * 1e9 + (t1.tv_nsec - t0.tv_nsec);
    double per_op_ns = elapsed_ns / n;
    printf("{\"experiment\":\"%s_perf\",\"library\":\"libnova\",", experiment);
    printf("\"count\":%d,\"total_ns\":%.0f,\"per_op_ns\":%.1f,", n, elapsed_ns, per_op_ns);
    printf("\"throughput_ops_s\":%.0f,\"_sink\":%.17e}\n",
           (double)n / (elapsed_ns * 1e-9), sink);
    free(jds);
}

/* Lane selector: LAB_LANE=geometric or LAB_LANE=geometric_vector routes to
 * the helio/geo rectangular path.  Anything else (unset, "apparent", legacy)
 * routes to the equ_coords apparent path.
 *
 * The orchestrator sends "geometric_vector" (the canonical lane token);
 * "geometric" is kept for backwards-compat with local invocations. */
int lane_is_geometric(void) {
    const char *lane = getenv("LAB_LANE");
    if (lane == NULL) return 0;
    return strcmp(lane, "geometric") == 0 || strcmp(lane, "geometric_vector") == 0;
}

void run_solar_position_perf(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    double *jds = malloc(n * sizeof(double));
    for (int i = 0; i < n; i++) {
        if (scanf("%lf", &jds[i]) != 1) {
            fprintf(stderr, "bad input line %d\n", i);
            exit(1);
        }
    }

    /* Warm-up: match functional scope (RA/Dec plus distance). */
    for (int i = 0; i < n && i < 100; i++) {
        struct ln_equ_posn equ;
        ln_get_solar_equ_coords(jds[i], &equ);
        double dist_au = ln_get_earth_solar_dist(jds[i]);
        (void)dist_au;
    }

    /* Timed run — perf contract: compute RA + Dec + distance */
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    double sink = 0.0;
    for (int i = 0; i < n; i++) {
        struct ln_equ_posn equ;
        ln_get_solar_equ_coords(jds[i], &equ);
        double dist_au = ln_get_earth_solar_dist(jds[i]);
        sink += equ.ra + equ.dec + dist_au;
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed_ns = (t1.tv_sec - t0.tv_sec) * 1e9 + (t1.tv_nsec - t0.tv_nsec);
    double per_op_ns = elapsed_ns / n;

    printf("{\"experiment\":\"solar_position_perf\",\"library\":\"libnova\",");
    printf("\"count\":%d,\"total_ns\":%.0f,\"per_op_ns\":%.1f,", n, elapsed_ns, per_op_ns);
    printf("\"throughput_ops_s\":%.0f,\"_sink\":%.17e}\n",
           (double)n / (elapsed_ns * 1e-9), sink);

    free(jds);
}

void run_lunar_position_perf(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    double *jds = malloc(n * sizeof(double));
    for (int i = 0; i < n; i++) {
        if (scanf("%lf", &jds[i]) != 1) {
            fprintf(stderr, "bad input line %d\n", i);
            exit(1);
        }
    }

    /* Warm-up: match functional scope (RA/Dec plus distance). */
    for (int i = 0; i < n && i < 100; i++) {
        struct ln_equ_posn equ;
        ln_get_lunar_equ_coords(jds[i], &equ);
        double dist_km = ln_get_lunar_earth_dist(jds[i]);
        (void)dist_km;
    }

    /* Timed run — perf contract: compute RA + Dec + distance */
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    double sink = 0.0;
    for (int i = 0; i < n; i++) {
        struct ln_equ_posn equ;
        ln_get_lunar_equ_coords(jds[i], &equ);
        double dist_km = ln_get_lunar_earth_dist(jds[i]);
        sink += equ.ra + equ.dec + dist_km;
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed_ns = (t1.tv_sec - t0.tv_sec) * 1e9 + (t1.tv_nsec - t0.tv_nsec);
    double per_op_ns = elapsed_ns / n;

    printf("{\"experiment\":\"lunar_position_perf\",\"library\":\"libnova\",");
    printf("\"count\":%d,\"total_ns\":%.0f,\"per_op_ns\":%.1f,", n, elapsed_ns, per_op_ns);
    printf("\"throughput_ops_s\":%.0f,\"_sink\":%.17e}\n",
           (double)n / (elapsed_ns * 1e-9), sink);

    free(jds);
}

void run_planet_position_perf(const char *experiment,
                                     planet_equ_fn_t equ_fn,
                                     planet_dist_fn_t dist_fn) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    double *jds = malloc(n * sizeof(double));
    for (int i = 0; i < n; i++) {
        if (scanf("%lf", &jds[i]) != 1) {
            fprintf(stderr, "bad input line %d\n", i);
            exit(1);
        }
    }

    for (int i = 0; i < n && i < 100; i++) {
        struct ln_equ_posn equ;
        equ_fn(jds[i], &equ);
        double dist_au = dist_fn(jds[i]);
        (void)dist_au;
    }

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    double sink = 0.0;
    for (int i = 0; i < n; i++) {
        struct ln_equ_posn equ;
        equ_fn(jds[i], &equ);
        double dist_au = dist_fn(jds[i]);
        sink += equ.ra + equ.dec + dist_au;
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed_ns = (t1.tv_sec - t0.tv_sec) * 1e9 + (t1.tv_nsec - t0.tv_nsec);
    double per_op_ns = elapsed_ns / n;

    printf("{\"experiment\":\"%s_perf\",\"library\":\"libnova\",", experiment);
    printf("\"count\":%d,\"total_ns\":%.0f,\"per_op_ns\":%.1f,", n, elapsed_ns, per_op_ns);
    printf("\"throughput_ops_s\":%.0f,\"_sink\":%.17e}\n",
           (double)n / (elapsed_ns * 1e-9), sink);

    free(jds);
}

File: pipeline/adapters/astropy_adapter/experiments/ephemerides.py
import json
import math
import sys
import time

try:
    from ..common import PLANET_EXPERIMENTS, PLANET_BARYCENTER_EXPERIMENTS, _astropy_geometric_geocentric
except ImportError:
    from common import PLANET_EXPERIMENTS, PLANET_BARYCENTER_EXPERIMENTS, _astropy_geometric_geocentric


def run_solar_position(lines_iter):
    """Sun geocentric RA/Dec via Astropy's public ephemeris API.

    ``ASTROPY_EPHEMERIS=jpl`` selects the JPL DE kernel (downloaded /
    cached by Astropy on first use); otherwise the bundled analytic
    ``builtin`` ephemeris is used.  Both go through
    ``solar_system_ephemeris.set(...) + get_body_barycentric`` so the
    published row honestly represents Astropy's public client API.
    """
    import os
    use_jpl = os.environ.get("ASTROPY_EPHEMERIS") == "jpl"
    ephemeris = "jpl" if use_jpl else "builtin"

    n = int(next(lines_iter).strip())
    cases = []
    for _ in range(n):
        jd_tt = float(next(lines_iter).strip())
        ra, dec, dist_au = _astropy_geometric_geocentric(jd_tt, "sun", ephemeris)
        cases.append({"jd_tt": jd_tt, "ra_rad": ra, "dec_rad": dec, "dist_au": dist_au})

    result = {
        "experiment": "solar_position",
        "library": "astropy_jpl" if use_jpl else "astropy",
        "model": (
            "Astropy solar_system_ephemeris('jpl') + get_body_barycentric"
            if use_jpl else "Astropy solar_system_ephemeris('builtin') + get_body_barycentric"
        ),
        "count": n,
        "cases": cases,
    }
    json.dump(result, sys.stdout, indent=None)
    print()


def run_lunar_position(lines_iter):
    """Moon geocentric RA/Dec via Astropy's public ephemeris API."""
    import os
    use_jpl = os.environ.get("ASTROPY_EPHEMERIS") == "jpl"
    ephemeris = "jpl" if use_jpl else "builtin"

    n = int(next(lines_iter).strip())
    cases = []
    for _ in range(n):
        jd_tt = float(next(lines_iter).strip())
        ra, dec, dist_au = _astropy_geometric_geocentric(jd_tt, "moon", ephemeris)
        cases.append({
            "jd_tt": jd_tt,
            "ra_rad": ra,
            "dec_rad": dec,
            "dist_km": dist_au * 149597870.700,
        })

    result = {
        "experiment": "lunar_position",
        "library": "astropy_jpl" if use_jpl else "astropy",
        "model": (
            "Astropy solar_system_ephemeris('jpl') + get_body_barycentric"
            if use_jpl else "Astropy solar_system_ephemeris('builtin') + get_body_barycentric"
        ),
        "count": n,
        "cases": cases,
    }
    json.dump(result, sys.stdout, indent=None)
    print()


def run_planet_position(lines_iter, experiment, planet_name, planet_np):
    """Planet geocentric RA/Dec via Astropy's public ephemeris API."""
    import os
    use_jpl = os.environ.get("ASTROPY_EPHEMERIS") == "jpl"
    ephemeris = "jpl" if use_jpl else "builtin"

    n = int(next(lines_iter).strip())
    cases = []
    for _ in range(n):
        jd_tt = float(next(lines_iter).strip())
        ra, dec, dist_au = _astropy_geometric_geocentric(jd_tt, planet_name.lower(), ephemeris)
        cases.append({"jd_tt": jd_tt, "ra_rad": ra, "dec_rad": dec, "dist_au": dist_au})

    result = {
        "experiment": experiment,
        "library": "astropy_jpl" if use_jpl else "astropy",
        "model": (
            f"Astropy solar_system_ephemeris('jpl') + get_body_barycentric({planet_name})"
            if use_jpl else f"Astropy solar_system_ephemeris('builtin') + get_body_barycentric({planet_name})"
        ),
        "count": n,
        "cases": cases,
    }
    json.dump(result, sys.stdout, indent=None)
    print()


def run_planet_barycenter_position(lines_iter, experiment, planet_name):
    """Planet system-barycenter geocentric RA/Dec using astropy JPL ephemeris.

    Uses ``get_body_barycentric(planet_name) - get_body_barycentric('earth')``
    which, with the JPL (DE440) ephemeris, returns the planet's system
    barycenter since DE440 stores planet barycenters (NAIF 4, 5, …) for
    Mars through Neptune.
    """
    n = int(next(lines_iter).strip())
    cases = []
    for _ in range(n):
        jd_tt = float(next(lines_iter).strip())
        ra, dec, dist_au = _astropy_geometric_geocentric(jd_tt, planet_name.lower(), "jpl")
        cases.append({"jd_tt": jd_tt, "ra_rad": ra, "dec_rad": dec, "dist_au": dist_au})

    result = {
        "experiment": experiment,
        "library": "astropy_jpl",
        "model": f"Astropy solar_system_ephemeris('jpl') + get_body_barycentric({planet_name}) [system barycenter]",
        "count": n,
        "cases": cases,
    }
    json.dump(result, sys.stdout, indent=None)
    print()


def run_solar_position_perf(lines_iter):
    """Performance measurement for solar position computation."""
    import os
    use_jpl = os.environ.get("ASTROPY_EPHEMERIS") == "jpl"
    ephemeris = "jpl" if use_jpl else "builtin"

    n = int(next(lines_iter).strip())
    jds = [float(next(lines_iter).strip()) for _ in range(n)]

    for i in range(min(n, 100)):
        _astropy_geometric_geocentric(jds[i], "sun", ephemeris)

    t0 = time.perf_counter_ns()
    sink = 0.0
    for jd in jds:
        ra, dec, dist = _astropy_geometric_geocentric(jd, "sun", ephemeris)
        sink += ra + dec + dist
    elapsed_ns = time.perf_counter_ns() - t0

    result = {
        "experiment": "solar_position_perf",
        "library": "astropy_jpl" if use_jpl else "astropy",
        "count": n,
        "total_ns": elapsed_ns,
        "per_op_ns": elapsed_ns / n,
        "throughput_ops_s": n / (elapsed_ns * 1e-9),
        "_sink": float(sink),
    }
    json.dump(result, sys.stdout, indent=None)
    print()


def run_lunar_position_perf(lines_iter):
    """Performance measurement for lunar position computation."""
    import os
    use_jpl = os.environ.get("ASTROPY_EPHEMERIS") == "jpl"
    ephemeris = "jpl" if use_jpl else "builtin"

    n = int(next(lines_iter).strip())
    jds = [float(next(lines_iter).strip()) for _ in range(n)]

    for i in range(min(n, 100)):
        _astropy_geometric_geocentric(jds[i], "moon", ephemeris)

    t0 = time.perf_counter_ns()
    sink = 0.0
    for jd in jds:
        ra, dec, dist_au = _astropy_geometric_geocentric(jd, "moon", ephemeris)
        sink += ra + dec + dist_au
    elapsed_ns = time.perf_counter_ns() - t0

    result = {
        "experiment": "lunar_position_perf",
        "library": "astropy_jpl" if use_jpl else "astropy",
        "count": n,
        "total_ns": elapsed_ns,
        "per_op_ns": elapsed_ns / n,
        "throughput_ops_s": n / (elapsed_ns * 1e-9),
        "_sink": float(sink),
    }
    json.dump(result, sys.stdout, indent=None)
    print()


def run_planet_position_perf(lines_iter, experiment, planet_np):
    """Performance measurement for planetary position computation."""
    import os
    use_jpl = os.environ.get("ASTROPY_EPHEMERIS") == "jpl"
    ephemeris = "jpl" if use_jpl else "builtin"
    planet_name = experiment.replace("_position", "").lower()

    n = int(next(lines_iter).strip())
    jds = [float(next(lines_iter).strip()) for _ in range(n)]

    for i in range(min(n, 100)):
        _astropy_geometric_geocentric(jds[i], planet_name, ephemeris)

    t0 = time.perf_counter_ns()
    sink = 0.0
    for jd in jds:
        ra, dec, dist_au = _astropy_geometric_geocentric(jd, planet_name, ephemeris)
        sink += ra + dec + dist_au
    elapsed_ns = time.perf_counter_ns() - t0

    result = {
        "experiment": f"{experiment}_perf",
        "library": "astropy_jpl" if use_jpl else "astropy",
        "count": n,
        "total_ns": elapsed_ns,
        "per_op_ns": elapsed_ns / n,
        "throughput_ops_s": n / (elapsed_ns * 1e-9),
        "_sink": float(sink),
    }
    json.dump(result, sys.stdout, indent=None)
    print()


def run_planet_barycenter_position_perf(lines_iter, experiment, planet_name):
    n = int(next(lines_iter).strip())
    jds = [float(next(lines_iter).strip()) for _ in range(n)]

    for i in range(min(n, 100)):
        _astropy_geometric_geocentric(jds[i], planet_name.lower(), "jpl")

    t0 = time.perf_counter_ns()
    sink = 0.0
    for jd in jds:
        ra, dec, dist = _astropy_geometric_geocentric(jd, planet_name.lower(), "jpl")
        sink += ra + dec + dist
    elapsed_ns = time.perf_counter_ns() - t0

    result = {
        "experiment": f"{experiment}_perf",
        "library": "astropy_jpl",
        "count": n,
        "total_ns": elapsed_ns,
        "per_op_ns": elapsed_ns / n,
        "throughput_ops_s": n / (elapsed_ns * 1e-9),
        "_sink": float(sink),
    }
    json.dump(result, sys.stdout, indent=None)
    print()
